File: examples/typescript/18_remote_scan_trace.ts
// (C) Copyright 2026- ECMWF and individual contributors.
//
// This software is licensed under the terms of the Apache Licence Version 2.0
// which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.

/**
 * Example 18 — Remote-scan tracing (TypeScript)
 *
 * Mirrors `examples/rust/src/bin/18_remote_scan_trace.rs`: stand up a
 * tiny Range-capable Node HTTP server, encode a multi-message `.tgm`,
 * then open it with `{ bidirectional: true, debug: true }` and watch
 * the dispatcher's `tensogram:scan:*` events.
 *
 * The dispatcher emits these events via `console.debug`.  The example
 * intercepts that channel so each event prints with a prefix; in real
 * code you'd forward them to whatever logging library you use.
 *
 * Run:
 *   cd typescript && npm run build
 *   cd ../examples/typescript && npm install
 *   npx tsx 18_remote_scan_trace.ts
 */

import { createServer, type IncomingMessage, type ServerResponse } from 'node:http';
import { encode, init, TensogramFile, type DataObjectDescriptor } from '@ecmwf.int/tensogram';

function describe(shape: number[]): DataObjectDescriptor {
  const strides = new Array<number>(shape.length).fill(1);
  for (let i = shape.length - 2; i >= 0; i--) strides[i] = strides[i + 1] * shape[i + 1];
  return {
    type: 'ntensor',
    ndim: shape.length,
    shape,
    strides,
    dtype: 'float32',
    byte_order: 'little',
    encoding: 'none',
    filter: 'none',
    compression: 'none',
  };
}

interface ServerHandle {
  url: string;
  close: () => Promise<void>;
}

function serveRangeBody(body: Uint8Array): Promise<ServerHandle> {
  const handler = (req: IncomingMessage, res: ServerResponse): void => {
    if (req.method === 'HEAD') {
      res.writeHead(200, {
        'Content-Length': body.byteLength,
        'Accept-Ranges': 'bytes',
      });
      res.end();
      return;
    }
    const range = req.headers['range'];
    if (typeof range === 'string' && range.startsWith('bytes=')) {
      const m = /^bytes=(\d+)?-(\d+)?$/.exec(range);
      if (!m) {
        res.writeHead(416);
        res.end();
        return;
      }
      const total = body.byteLength;
      let start: number;
      let end: number;
      if (m[1] === undefined && m[2] !== undefined) {
        const suffix = parseInt(m[2], 10);
        start = Math.max(0, total - suffix);
        end = total - 1;
      } else {
        start = parseInt(m[1] ?? '0', 10);
        end = m[2] !== undefined ? parseInt(m[2], 10) : total - 1;
      }
      const slice = Buffer.from(body.buffer, body.byteOffset + start, end - start + 1);
      res.writeHead(206, {
        'Content-Range': `bytes ${start}-${end}/${total}`,
        'Content-Length': slice.byteLength,
      });
      res.end(slice);
      return;
    }
    res.writeHead(200, { 'Content-Length': body.byteLength });
    res.end(Buffer.from(body));
  };
  return new Promise<ServerHandle>((resolve) => {
    const server = createServer(handler);
    server.listen(0, '127.0.0.1', () => {
      const addr = server.address();
      if (!addr || typeof addr === 'string') {
        throw new Error('server.address() returned an unexpected shape');
      }
      const url = `http://127.0.0.1:${addr.port}/forecast.tgm`;
      resolve({
        url,
        close: () => new Promise<void>((res) => server.close(() => res())),
      });
    });
  });
}

async function main(): Promise<void> {
  await init();

  const N = 5;
  const messages: Uint8Array[] = [];
  for (let i = 0; i < N; i++) {
    const data = new Float32Array(16).map((_, j) => i * 100 + j);
    messages.push(
      encode(
        { version: 3, base: [{ msg_index: i }] },
        [{ descriptor: describe([16]), data }],
      ),
    );
  }
  const total = messages.reduce((acc, m) => acc + m.byteLength, 0);
  const file = new Uint8Array(total);
  let offset = 0;
  for (const m of messages) {
    file.set(m, offset);
    offset += m.byteLength;
  }
  console.log(`Encoded ${N} messages, ${file.byteLength} bytes total`);

  const server = await serveRangeBody(file);
  const originalDebug = console.debug.bind(console);
  console.debug = (label: unknown, payload?: unknown) => {
    if (typeof label === 'string' && label.startsWith('tensogram:')) {
      originalDebug(`  [trace] ${label}`, payload ?? '');
    } else {
      originalDebug(label as string, payload as object);
    }
  };

  try {
    console.log(`\nServing at ${server.url}`);
    console.log('\n── Forward-only walker (default) ──');
    const fwd = await TensogramFile.fromUrl(server.url, { debug: true });
    try {
      console.log(`Opened: ${fwd.messageCount} messages`);
    } finally {
      fwd.close();
    }

    console.log('\n── Bidirectional walker (opt-in) ──');
    const bid = await TensogramFile.fromUrl(server.url, {
      bidirectional: true,
      debug: true,
    });
    try {
      console.log(`Opened: ${bid.messageCount} messages`);
    } finally {
      bid.close();
    }

    console.log('\nDispatcher events streamed via console.debug above.');
    console.log('Replace the interceptor with your own logger to capture them.');
  } finally {
    console.debug = originalDebug;
    await server.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/rust/src/bin/18_remote_scan_trace.rs
// (C) Copyright 2026- ECMWF and individual contributors.
//
// This software is licensed under the terms of the Apache Licence Version 2.0
// which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
// In applying this licence, ECMWF does not waive the privileges and immunities
// granted to it by virtue of its status as an intergovernmental organisation nor
// does it submit to any jurisdiction.

//! Example 18 — Remote-Scan Tracing
//!
//! Subscribes to the `tensogram::remote_scan` tracing target and runs
//! a full-file iteration with the bidirectional walker enabled, so every
//! mode/round/footer-eager event surfaces on stderr.  The subscriber
//! `EnvFilter` is configured in code (no `RUST_LOG` env-var required);
//! point your environment at it if you want a different filter.
//!
//! Run with:
//!   cargo run -p tensogram-rust-examples --features remote --bin 18_remote_scan_trace

use std::io::{Read, Write};
use std::net::TcpListener;
use std::sync::Arc;

use tensogram::{
    ByteOrder, DataObjectDescriptor, DecodeOptions, Dtype, EncodeOptions, GlobalMetadata,
    RemoteScanOptions, TensogramFile,
};
use tracing_subscriber::EnvFilter;

const N_MESSAGES: usize = 5;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let filter = EnvFilter::try_from_default_env()
        .unwrap_or_else(|_| EnvFilter::new("tensogram::remote_scan=debug,tensogram::remote=info"));
    tracing_subscriber::fmt()
        .with_env_filter(filter)
        .with_target(true)
        .with_writer(std::io::stderr)
        .init();

    let mut file_bytes = Vec::new();
    for i in 0..N_MESSAGES {
        let meta = GlobalMetadata::default();
        let desc = DataObjectDescriptor {
            obj_type: "ntensor".to_string(),
            ndim: 1,
            shape: vec![4],
            strides: vec![1],
            dtype: Dtype::Float32,
            byte_order: ByteOrder::native(),
            encoding: "none".to_string(),
            filter: "none".to_string(),
            compression: "none".to_string(),
            params: std::collections::BTreeMap::new(),
            masks: None,
        };
        let payload: Vec<u8> = (0..16).map(|b| b as u8 + i as u8).collect();
        let msg =
            tensogram::encode::encode(&meta, &[(&desc, &payload)], &EncodeOptions::default())?;
        file_bytes.extend_from_slice(&msg);
    }
    println!(
        "Encoded {N_MESSAGES} messages, {} bytes total",
        file_bytes.len()
    );

    let tgm = Arc::new(file_bytes);
    let listener = TcpListener::bind("127.0.0.1:0")?;
    let port = listener.local_addr()?.port();
    let url = format!("http://127.0.0.1:{port}/forecast.tgm");
    let server_data = tgm.clone();
    std::thread::spawn(move || serve_http(listener, &server_data));

    println!("Serving at {url}");
    println!("\n── Forward-only walker (default) ──");
    let storage = std::collections::BTreeMap::new();
    let file = TensogramFile::open_remote(&url, &storage, None)?;
    println!("Opened: {} messages", file.message_count()?);
    for i in 0..N_MESSAGES {
        let _ = file.decode_object(i, 0, &DecodeOptions::default())?;
    }
    drop(file);

    println!("\n── Bidirectional walker (opt-in) ──");
    let bidi_opts = Some(RemoteScanOptions {
        bidirectional: true,
    });
    let file = TensogramFile::open_remote(&url, &storage, bidi_opts)?;
    println!("Opened: {} messages", file.message_count()?);
    for i in 0..N_MESSAGES {
        let _ = file.decode_object(i, 0, &DecodeOptions::default())?;
    }

    println!("\nTraced events emitted to stderr.  Re-run with");
    println!("`RUST_LOG=tensogram::remote_scan=trace` for finer granularity.");
    Ok(())
}

fn serve_http(listener: TcpListener, data: &[u8]) {
    for stream in listener.incoming() {
        let mut stream = match stream {
            Ok(s) => s,
            Err(_) => continue,
        };
        let mut buf = vec![0u8; 4096];
        let n = match stream.read(&mut buf) {
            Ok(n) => n,
            Err(_) => continue,
        };
        let request = String::from_utf8_lossy(&buf[..n]);
        let is_head = request.starts_with("HEAD ");
        let range = parse_range(&request, data.len());
        match range {
            Some((start, end)) => {
                let chunk = &data[start..end];
                let header = format!(
                    "HTTP/1.1 206 Partial Content\r\nContent-Range: bytes {start}-{}/{}\r\nContent-Length: {}\r\nAccept-Ranges: bytes\r\nConnection: close\r\n\r\n",
                    end - 1,
                    data.len(),
                    chunk.len(),
                );
                let _ = stream.write_all(header.as_bytes());
                if !is_head {
                    let _ = stream.write_all(chunk);
                }
            }
            None => {
                let header = format!(
                    "HTTP/1.1 200 OK\r\nContent-Length: {}\r\nAccept-Ranges: bytes\r\nConnection: close\r\n\r\n",
                    data.len(),
                );
                let _ = stream.write_all(header.as_bytes());
                if !is_head {
                    let _ = stream.write_all(data);
                }
            }
        }
    }
}

fn parse_range(request: &str, file_len: usize) -> Option<(usize, usize)> {
    let line = request
        .lines()
        .find(|l| l.to_ascii_lowercase().starts_with("range:"))?;
    let spec = line.split("bytes=").nth(1)?.trim();
    if let Some(suffix) = spec.strip_prefix('-') {
        let n: usize = suffix.parse().ok()?;
        Some((file_len.saturating_sub(n), file_len))
    } else {
        let mut parts = spec.split('-');
        let start: usize = parts.next()?.parse().ok()?;
        let end_str = parts.next()?;
        let end = if end_str.is_empty() {
            file_len
        } else {
            end_str.parse::<usize>().ok()? + 1
        };
        Some((start, end.min(file_len)))
    }
}
